### ISSUE TITLE
scikit learn decision tree gets unpickled as dictionary

### ISSUE THREAD

**Issue by nikos-daniilidis** [jsonpickle/jsonpickle#155]
I am having a problem unpickling scikit-learn deicsion trees. Example code:

```python
import numpy as np
import jsonpickle
import jsonpickle.ext.numpy as jsonpickle_numpy
from sklearn.tree import DecisionTreeClassifier

#  create data
np.random.seed(13)
x = np.random.randint(low=0, high = 10, size = 12)
X = x.reshape(4, 3)
Y = np.random.randint(low=0, high=2, size=4)
Y.reshape(-1,1)

# train model
clf = DecisionTreeClassifier(max_depth=1)
clf.fit(X, Y)

# freeze and thaw
jsonpickle_numpy.register_handlers()
p = jsonpickle.pickler.Pickler()
u = jsonpickle.unpickler.Unpickler()
j = p.flatten(clf)
clf1 = u.restore(json.loads(json.dumps(p.flatten(clf))))

# predict from thawed
xx = np.array([1, 2, 3])
xx = xx.reshape(1,-1)
clf1.predict(xx)

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-15-4e5a1309ca7a> in <module>()
     27 xx = np.array([1, 2, 3])
     28 xx = xx.reshape(1,-1)
---> 29 clf1.predict(xx)

/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sklearn/tree/tree.pyc in predict(self, X, check_input)
    403 
    404         X = self._validate_X_predict(X, check_input)
--> 405         proba = self.tree_.predict(X)
    406         n_samples = X.shape[0]
    407 

AttributeError: 'dict' object has no attribute 'predict'
```

On further inspection, the problem seems to be the following:

```python
print type(clf)
print type(clf1)
print type(clf.tree_)
print type(clf1.tree_)

<class 'sklearn.tree.tree.DecisionTreeClassifier'>
<class 'sklearn.tree.tree.DecisionTreeClassifier'>
<type 'sklearn.tree._tree.Tree'>
<type 'dict'>
```

i.e. the classifier object gets unpickled as the correct type, while the decision tree under `clf.tree_` is getting unpickled as a dictionary. I confirned the same behavior with `DecisionTreeRegressor` and ensemble methods which use trees , for example `GradientBoostingClassifier`.

**Comment by mariolito** [comment]
Hi all,

I have exactly the same issue.

Still no suggestions ?

Thanks,

Marios

**Comment by marcintustin** [comment]
Would you mind posting the faulty JSON output and also the code used to generate it? That would be illuminating. The most minimal example possible would be highly preferable.

**Comment by marcintustin** [comment]
This looks pretty big for an MWE. Is all of it necessary?

**Comment by nikos-daniilidis** [comment]
Posting a smaller example. Not sure I can make it smaller than this: 

`import numpy as np
import json
import jsonpickle
import jsonpickle.ext.numpy as jsonpickle_numpy
from sklearn.tree import DecisionTreeClassifier

#  create data
X = np.array([0, 1, 2, 3])
Y = np.array([0, 0, 1, 1])
X = X.reshape(4, 1)
Y = Y.reshape(-1,1)

clf = DecisionTreeClassifier(max_depth=1)
clf.fit(X, Y)

# freeze and thaw
jsonpickle_numpy.register_handlers()
p = jsonpickle.pickler.Pickler()
u = jsonpickle.unpickler.Unpickler()
j = p.flatten(clf)
clf1 = u.restore(json.loads(json.dumps(p.flatten(clf))))

# predict from thawed
xx = np.array([1, 2, 3])
xx = xx.reshape(3,1)
clf1.predict(xx)`

I get a similar `ValueError` as above. The contents of the `j` JSON are: 

`{
    "py/object": "sklearn.tree.tree.DecisionTreeClassifier", 
    "py/state": {
        "presort": false, 
        "splitter": "best", 
        "tree_": {
            "py/object": "sklearn.tree._tree.Tree", 
            "py/state": {
                "node_count": 3, 
                "nodes": {
                    "py/object": "numpy.ndarray", 
                    "dtype": "[('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]", 
                    "values": [
                        {
                            "py/tuple": [
                                1, 
                                2, 
                                0, 
                                1.5, 
                                0.5, 
                                4, 
                                4.0
                            ]
                        }, 
                        {
                            "py/tuple": [
                                -1, 
                                -1, 
                                -2, 
                                -2.0, 
                                0.0, 
                                2, 
                                2.0
                            ]
                        }, 
                        {
                            "py/tuple": [
                                -1, 
                                -1, 
                                -2, 
                                -2.0, 
                                0.0, 
                                2, 
                                2.0
                            ]
                        }
                    ]
                }, 
                "values": {
                    "py/object": "numpy.ndarray", 
                    "dtype": "float64", 
                    "values": [
                        [
                            [
                                2.0, 
                                2.0
                            ]
                        ], 
                        [
                            [
                                2.0, 
                                0.0
                            ]
                        ], 
                        [
                            [
                                0.0, 
                                2.0
                            ]
                        ]
                    ]
                }, 
                "max_depth": 1
            }
        }, 
        "n_features_": 1, 
        "n_outputs_": 1, 
        "n_classes_": {
            "py/object": "numpy.int64", 
            "dtype": "int64", 
            "value": 2
        }, 
        "max_leaf_nodes": null, 
        "classes_": {
            "py/object": "numpy.ndarray", 
            "dtype": "int64", 
            "values": [
                0, 
                1
            ]
        }, 
        "max_features_": 1, 
        "min_samples_leaf": 1, 
        "min_samples_split": 2, 
        "min_weight_fraction_leaf": 0.0, 
        "criterion": "gini", 
        "random_state": null, 
        "min_impurity_split": 1e-07, 
        "max_features": null, 
        "_sklearn_version": "0.18.1", 
        "max_depth": 1, 
        "class_weight": null
    }
}`

Thanks for looking into this!

**Comment by marcintustin** [comment]
Hmm. I'm on mobile so perhaps I'm misreading but (a) I don't see a value
error anywhere; (b) the attribute error I do see posted refers to a dict.
In the posted json the type of tree_ is not dict.

Please post the exact stack trace you get, or confirm that the exception
you see is the attribute error in the very first post in this issue.

Assuming you are getting that attribute error then the issue is on the
unpickling side. Given that, I would love to see what happens when you
unpickle just the tree_ attribute.

On Wed, Jan 25, 2017 at 9:54 PM Nikos Daniilidis <notifications@github.com>
wrote:

Posting a smaller example. Not sure I can make it smaller than this:

`import numpy as np


import json
import jsonpickle
import jsonpickle.ext.numpy as jsonpickle_numpy
from sklearn.tree import DecisionTreeClassifier

create data

X = np.array([0, 1, 2, 3])
Y = np.array([0, 0, 1, 1])
X = X.reshape(4, 1)
Y = Y.reshape(-1,1)

clf = DecisionTreeClassifier(max_depth=1)
clf.fit(X, Y)
freeze and thaw

jsonpickle_numpy.register_handlers()
p = jsonpickle.pickler.Pickler()
u = jsonpickle.unpickler.Unpickler()
j = p.flatten(clf)
clf1 = u.restore(json.loads(json.dumps(p.flatten(clf))))
predict from thawed

xx = np.array([1, 2, 3])

xx = xx.reshape(3,1)
clf1.predict(xx)`

I get a similar ValueError as above. The contents of the j JSON are:

{ "py/object": "sklearn.tree.tree.DecisionTreeClassifier", "py/state": {
"presort": false, "splitter": "best", "tree_": { "py/object":
"sklearn.tree._tree.Tree", "py/state": { "node_count": 3, "nodes": {
"py/object": "numpy.ndarray", "dtype": "[('left_child', '<i8'),
('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'),
('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples',
'<f8')]", "values": [ { "py/tuple": [ 1, 2, 0, 1.5, 0.5, 4, 4.0 ] }, {
"py/tuple": [ -1, -1, -2, -2.0, 0.0, 2, 2.0 ] }, { "py/tuple": [ -1, -1,
-2, -2.0, 0.0, 2, 2.0 ] } ] }, "values": { "py/object": "numpy.ndarray",
"dtype": "float64", "values": [ [ [ 2.0, 2.0 ] ], [ [ 2.0, 0.0 ] ], [ [
0.0, 2.0 ] ] ] }, "max_depth": 1 } }, "n_features_": 1, "n_outputs_": 1,
"n_classes_": { "py/object": "numpy.int64", "dtype": "int64", "value": 2 },
"max_leaf_nodes": null, "classes_": { "py/object": "numpy.ndarray",
"dtype": "int64", "values": [ 0, 1 ] }, "max_features_": 1,
"min_samples_leaf": 1, "min_samples_split": 2, "min_weight_fraction_leaf":
0.0, "criterion": "gini", "random_state": null, "min_impurity_split":
1e-07, "max_features": null, "_sklearn_version": "0.18.1", "max_depth": 1,
"class_weight": null } }

Thanks for looking into this!

—
You are receiving this because you commented.


Reply to this email directly, view it on GitHub
<https://github.com/jsonpickle/jsonpickle/issues/155#issuecomment-275296757>,
or mute the thread

**Comment by nikos-daniilidis** [comment]
Sorry, that was my bad. The error trace from the latter example is: 

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/sklearn/tree/tree.py", line 405, in predict
    proba = self.tree_.predict(X)
AttributeError: 'dict' object has no attribute 'predict'
```

And, indeed, the problem is that the `tree_` attribute is getting unpickled as a dict (see also comments on my original post in the thread):

```
print type(clf1.tree_)
<type 'dict'>
```

**Comment by nikos-daniilidis** [comment]
And you are correct about the JSON, I mistakenly posted the result of flattening the classifier itself. After thawing, the JSON is as follows: 

```
j = p.flatten(clf1)
print json.dumps(j, indent=4)

{
    "py/object": "sklearn.tree.tree.DecisionTreeClassifier", 
    "py/state": {
        "presort": false, 
        "splitter": "best", 
        "tree_": {}, 
        "n_features_": 1, 
        "n_outputs_": 1, 
        "n_classes_": {
            "py/object": "numpy.int64", 
            "dtype": "int64", 
            "value": 2
        }, 
        "max_leaf_nodes": null, 
        "classes_": {
            "py/object": "numpy.ndarray", 
            "dtype": "int64", 
            "values": [
                0, 
                1
            ]
        }, 
        "max_features_": 1, 
        "min_samples_leaf": 1, 
        "min_samples_split": 2, 
        "min_weight_fraction_leaf": 0.0, 
        "criterion": "gini", 
        "random_state": null, 
        "min_impurity_split": 1e-07, 
        "max_features": null, 
        "_sklearn_version": "0.18.1", 
        "max_depth": 1, 
        "class_weight": null
    }
}
```

The tree is an empty dict, it seems.

**Comment by marcintustin** [comment]
Ah that's useful, and makes more sense. Can you post what the type of `tree_` is and what its state is at the point of taking that pickle? That will really help zero in on the funkiness.

**Comment by nikos-daniilidis** [comment]
Sure. These are immediately after the `clf1 = u.restore(json.loads(json.dumps(p.flatten(clf))))` line: 

```
>>> print type(clf1.tree_)
<type 'dict'>
```

and the state is:

```
>>> print clf1.tree_
{u'py/object': u'sklearn.tree._tree.Tree', u'py/state': {u'node_count': 3, u'values': {u'py/object': u'numpy.ndarray', u'dtype': u'float64', u'values': [[[2.0, 2.0]], [[2.0, 0.0]], [[0.0, 2.0]]]}, u'nodes': {u'py/object': u'numpy.ndarray', u'dtype': u"[('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]", u'values': [{u'py/tuple': [1, 2, 0, 1.5, 0.5, 4, 4.0]}, {u'py/tuple': [-1, -1, -2, -2.0, 0.0, 2, 2.0]}, {u'py/tuple': [-1, -1, -2, -2.0, 0.0, 2, 2.0]}]}, u'max_depth': 1}}
```

**Comment by marcintustin** [comment]
@nikos-daniilidis I'm interested in type and state *before* it is pickled, not after it is unpickled.

**Comment by nikos-daniilidis** [comment]
Before pickling: 

```
>>> print type(clf.tree_)
<type 'sklearn.tree._tree.Tree'>
```

and 

```
>>> print clf.tree_
<sklearn.tree._tree.Tree object at 0x10697a3d8>

>>> print (clf.tree_.__doc__)

Array-based representation of a binary decision tree.

    The binary tree is represented as a number of parallel arrays. The i-th
    element of each array holds information about the node `i`. Node 0 is the
    tree's root. You can find a detailed description of all arrays in
    `_tree.pxd`. NOTE: Some of the arrays only apply to either leaves or split
    nodes, resp. In this case the values of nodes of the other type are
    arbitrary!

    Attributes
    ----------
    node_count : int
        The number of nodes (internal nodes + leaves) in the tree.

    capacity : int
        The current capacity (i.e., size) of the arrays, which is at least as
        great as `node_count`.

    max_depth : int
        The maximal depth of the tree.

    children_left : array of int, shape [node_count]
        children_left[i] holds the node id of the left child of node i.
        For leaves, children_left[i] == TREE_LEAF. Otherwise,
        children_left[i] > i. This child handles the case where
        X[:, feature[i]] <= threshold[i].

    children_right : array of int, shape [node_count]
        children_right[i] holds the node id of the right child of node i.
        For leaves, children_right[i] == TREE_LEAF. Otherwise,
        children_right[i] > i. This child handles the case where
        X[:, feature[i]] > threshold[i].

    feature : array of int, shape [node_count]
        feature[i] holds the feature to split on, for the internal node i.

    threshold : array of double, shape [node_count]
        threshold[i] holds the threshold for the internal node i.

    value : array of double, shape [node_count, n_outputs, max_n_classes]
        Contains the constant prediction value of each node.

    impurity : array of double, shape [node_count]
        impurity[i] holds the impurity (i.e., the value of the splitting
        criterion) at node i.

    n_node_samples : array of int, shape [node_count]
        n_node_samples[i] holds the number of training samples reaching node i.

    weighted_n_node_samples : array of int, shape [node_count]
        weighted_n_node_samples[i] holds the weighted number of training samples
        reaching node i.
```

Not sure what else would be helpful here, feel free to ask for specific attributes!

**Comment by marcintustin** [comment]
@nikos-daniilidis Can you also `print (clf.tree_.__dict__)` so I can see the state? Or whatever else is appropriate in case this is using a state tuple or something more exotic.

**Comment by nikos-daniilidis** [comment]
The state seems to be stored in the `__getstate__` method. 

```
>>> clf.tree_.__getstate__()
{'node_count': 3, 'nodes': array([(1, 2, 0, 1.5, 0.5, 4, 4.0), (-1, -1, -2, -2.0, 0.0, 2, 2.0),
       (-1, -1, -2, -2.0, 0.0, 2, 2.0)], 
      dtype=[('left_child', '<i8'), ('right_child', '<i8'), ('feature', '<i8'), ('threshold', '<f8'), ('impurity', '<f8'), ('n_node_samples', '<i8'), ('weighted_n_node_samples', '<f8')]), 'values': array([[[ 2.,  2.]],

       [[ 2.,  0.]],

       [[ 0.,  2.]]]), 'max_depth': 1}
```

The other attributes I find are: 

```
>>> print clf.tree_.__class__
<type 'sklearn.tree._tree.Tree'>
>>> print clf.tree_.__delattr__
<method-wrapper '__delattr__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__format__
<built-in method __format__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__getattribute__
<method-wrapper '__getattribute__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__getstate__
<built-in method __getstate__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__hash__
<method-wrapper '__hash__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__init__
<method-wrapper '__init__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__new__
<built-in method __new__ of type object at 0x105a8e450>
>>> print clf.tree_.__pyx_vtable__
<capsule object NULL at 0x1059e90f0>
>>> print clf.tree_.__reduce__
<built-in method __reduce__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__reduce_ex__
<built-in method __reduce_ex__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__repr__
<method-wrapper '__repr__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__setattr__
<method-wrapper '__setattr__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__setstate__
<built-in method __setstate__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__sizeof__
<built-in method __sizeof__ of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__str__
<method-wrapper '__str__' of sklearn.tree._tree.Tree object at 0x105da4648>
>>> print clf.tree_.__subclasshook__
<built-in method __subclasshook__ of type object at 0x105a8e450>
```

**Comment by marcintustin** [comment]
Perfect, thank you!

**Comment by thoraxe** [comment]
What was the resolution to this issue?

**Comment by davvid** [comment]
This was originally fixed in #170 which was first released in jsonpickle v0.9.5 (July 2017), but I just added a test case to ensure that it works forevermore.  We're currently at v1.5.0.
